### PR TITLE
Switch the system to RecyclerView and fix the screen rotation

### PR DESCRIPTION
**Pull request by Jimvy**
Landscape mode works now #6 .

Advantages of RecyclerView:
- No need to worry about placing the new sound buttons in the layouts, the RecyclerView's LayoutManager does it for you!
- No need to copy-paste button layouts either
- No need to create an `@+id/` that matches the `@string/`.
- In theory it should be faster to render as the RecyclerView can recycle the views + the layouts are WAY smaller (lower depth) so the system has an easier task.

Disadvantages of RecyclerView:
- The Kotlin code size has increased.
- The generation of the sounds list is done at fragment creation, so maybe it slows down infinitesimally the startup.

I haven't checked the performance however, as my phone is a bit too fast to detect a significant difference.

Things that need improvement:
- Adding a sound to the list requires adding it to the list of sounds in data/Sound.kt (in addition to strings.xml and to the res/raw folder). This could be improved by either auto-scanning the res/raw folder, or by using a DB-like file (I'm thinking of a simple JSON or YAML file right now).
- For now the list is not sorted automatically. I think it is the easiest improvement to make.

Things that may be improved:
- The number of columns can probably be determined by the horizontal width of the screen. So instead of referring to smallest-width+portrait/landscape, it should be possible to refer to width, and use files in values-w600dp, values-840dp, values-1000dp (or other, need to check what are the actual sizes).
- The random sound button is pinned on top of the screen in every layout. An alternative would be to include it in the RecyclerView when the height is too small, like in landscape mode on phone. This could be done by having a layout/fragment_main.xml for small heights (random button in RV and not in layout), and a layout-h400dp/fragment_main.xml for non-small heights (random button in layout).
- Animating the button is a bit more complex: it doesn't use an ObjectAnimator with a target anymore; rather, it uses a ValueAnimator and updates the values itself. So changing the animation will be a tad more difficult.

Issues that still exist:
- Stopping a sound doesn't stop the animation; may be fixed by refactoring the Discotheque
- In sequential mode, no animation is displayed, because the Discotheque returns 0. Again, this could be solved by making Discotheque accept start/stop/pause/resume listeners.
- Random button in simultaneous mode: suppose that the first tap starts a long sound, and a second tap starts a shorter sound, such that the second sound will finish before the first one, then one sees first the animation for the 2nd sound, and when finished, it sees the (partially completed) animation for the 1st sound. Not really a problem though.
- The layout of the items: the margin appear on top and on bottom of it, so there is 2x the margin between the items and 1x the margin at the top/bottom of the list. It looks OK though.

### CODE REVIEW

File: app/build.gradle
Comment (TGLuis) [inline app/build.gradle]: What is the point of this line if it is not needed yet? What could it be needed for?

File: app/src/main/java/data/Sound.kt
Comment (TGLuis) [inline app/src/main/java/data/Sound.kt]: Is there a reason not to have done: 
`return 31 * id + name.hashCode()` ?
why multiply the id by 31?

File: app/src/main/java/fragments/MainFragment.kt
Comment (TGLuis) [inline app/src/main/java/fragments/MainFragment.kt:46]: I don't think there will be any difference for this case.

File: app/src/main/res/layout/sound_button.xml
Comment (TGLuis) [inline app/src/main/res/layout/sound_button.xml:7]: In fact, I was about to say exactly the same :+1:

File: app/src/main/res/layout/sound_button.xml
Comment (TGLuis) [inline app/src/main/res/layout/sound_button.xml]: I don't think we should let that here if it does not have any purpose...

File: app/src/main/java/fragments/MainFragment.kt
Comment (TGLuis) [inline app/src/main/java/fragments/MainFragment.kt]: I don't understand this comment, why is it needed to add those?

File: app/src/main/java/data/Sound.kt
Comment (TGLuis) [inline app/src/main/java/data/Sound.kt]: `return id == other.id && name == other.name` ?

File: app/src/main/res/layout/sound_button.xml
Comment (TGLuis) [inline app/src/main/res/layout/sound_button.xml:13]: The margin should change with the size of the layout, I think...

File: app/src/main/java/fragments/MainFragment.kt
Comment (TGLuis) [inline app/src/main/java/fragments/MainFragment.kt]: why this object is not made like the animCleanup?  To be consistent, shouldn't it be better if we put it as a val like the other object?

File: app/src/main/java/adapter/SoundAdapter.kt
Comment (TGLuis) [inline app/src/main/java/adapter/SoundAdapter.kt]: Are those functions really needed? They do not seem to be called anywhere...

**Review by TGLuis — COMMENTED** [review]
It seems to be a really good PR, but some improvements can be added and some questions need answers. It uses dark functionalities (or magic?) of Kotlin and some good tricks but it is not easy to grasp at a first look. Could be best with some comment. 
I wait for the answers of my questions and will approve if it is satisfactory! :100:

File: app/build.gradle
Comment (Jimvy) [reply app/build.gradle]: There is no point in this line. Will be removed

File: app/src/main/java/data/Sound.kt
Comment (Jimvy) [reply app/src/main/java/data/Sound.kt]: This was generated automatically by IntelliJ. I don't really know why ^^

File: app/src/main/java/fragments/MainFragment.kt
Comment (Jimvy) [reply app/src/main/java/fragments/MainFragment.kt:46]: Actually, why is there two different methods anyway?
In the original version onCreateView inflated the layout, and onActivityCreated attached the listeners to the buttons and set the menu.
What kind of features cannot be put in onCreateView and need to be put in onActivityCreated, and what kind of features should not be put in onActivityCreated and should be put in onCreateView?
I would be glad to hear your opinion ^^

File: app/src/main/res/layout/sound_button.xml
Comment (Jimvy) [reply app/src/main/res/layout/sound_button.xml:7]: Simple solution: remove the string altogether ^^

File: app/src/main/java/fragments/MainFragment.kt
Comment (Jimvy) [reply app/src/main/java/fragments/MainFragment.kt]: This is for myself for the search feature

File: app/src/main/java/data/Sound.kt
Comment (Jimvy) [reply app/src/main/java/data/Sound.kt]: IntelliJ is not smart enough to figure this out

File: app/src/main/res/layout/sound_button.xml
Comment (Jimvy) [reply app/src/main/res/layout/sound_button.xml:13]: This one will be more complicated to fix than I expected...

File: app/src/main/java/fragments/MainFragment.kt
Comment (Jimvy) [reply app/src/main/java/fragments/MainFragment.kt]: Yep, it would be cleaner

File: app/src/main/java/adapter/SoundAdapter.kt
Comment (Jimvy) [reply app/src/main/java/adapter/SoundAdapter.kt]: They will be used for the search feature, but yes, they are unused.

File: app/src/main/java/fragments/MainFragment.kt
Comment (TGLuis) [reply app/src/main/java/fragments/MainFragment.kt:46]: I would say: everything that is static related in onCreateView and everything that is dynamic like onClickListeners etc, in onActivityCreated but I am not sure I followed this line of conduct all the time myself :p

File: app/src/main/res/layout/sound_button.xml
Comment (TGLuis) [reply app/src/main/res/layout/sound_button.xml:13]: Should I wait for you to fix this before merging or you want to clean this before the merge?

Comment (Jimvy) [reply app/src/main/res/layout/sound_button.xml:13]: I'll wait until @FrancoisDuchene has tested the layouts to inform if it is necessary.

File: app/src/main/java/adapter/SoundAdapter.kt
Comment (FrancoisDuchene) [inline app/src/main/java/adapter/SoundAdapter.kt]: From line 83 -> 98, it is normal that every method to add/remove sounds to `soundsList` have been removed ?

**Review by FrancoisDuchene — COMMENTED** [review]
Good PR :+1: Only one trivial question then i can approve !
The system to add sounds is pretty well made, the solution to add a JSON file is a good choice for the future. (Using a database like SQLite might be overkill for this app); It would be useful to make only one or two changes to add sounds. For example, add the filename in the uppermentioned **JSON** file and the sound in the *raw* folder.

(R.I.P. my poor handcrafted layouts :cry: )

**Review by FrancoisDuchene — APPROVED** [review]

File: app/src/main/java/adapter/SoundAdapter.kt
Comment (Jimvy) [reply app/src/main/java/adapter/SoundAdapter.kt]: They are useless (they could be used if we decide to let the user add/remove sounds to the list, or if we add/remove the random sound button to the list dynamically (ex: in landscape mode if the height is not large enough)

**Pull request merged**
